HTTP library that uses net-app

jukkar opened the pull request:
This is a proposal for new HTTP library that uses net-app API instead of running things directly on top of net_context.
This supports both HTTP client and server functionality, both are also supporting TLS. Old HTTP API is deprecated.
I will also send a PR that will add websocket support to this new API.

jukkar committed:
net: http: Create HTTP library that uses net-app

Create http library that uses net-app instead of net_context
directly. The old HTTP API is deprecated.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: HTTP server sample using net-app

This is a sample HTTP server application that uses new http API
instead of legacy one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

lpereira commented on subsys/net/lib/http/http_app_server.c:
```diff
@@ -0,0 +1,1085 @@
+			 ntohs(net_sin(addr)->sin_port));
+#endif
+	} else {
+		snprintk(buf, buflen, "<invalid address family %d>",
```
Although this won't crash, if buflen=NET_IPV4_ADDR_LEN, this will truncate this string if the address family is invalid.

jukkar replied:
Sure, not a big issue thou as this is just used for debug printing.

lpereira commented on subsys/net/lib/http/http_app_server.c:
```diff
@@ -0,0 +1,1085 @@
+}
+
+#if defined(CONFIG_HTTPS)
+int http_server_set_tls(struct http_ctx *ctx,
```
A future improvement could redirect all HTTP requests to HTTPS if TLS is enabled.

jukkar replied:
Sure, but this depends on how application is doing things. If we have enabled HTTPS, then HTTP port is not normally listened at all. So application would need to create a listener to port 80 and then re-direct traffic to HTTPS port in this case.

lpereira commented on subsys/net/lib/http/http_app_server.c:
```diff
@@ -0,0 +1,1085 @@
+	}
+}
+
+struct http_root_url *http_server_add_url(struct http_server_urls *my,
```
If the URLs array is sorted by URL length, it's possible to quickly perform a longest subprefix match without requiring fancier data structures like a trie.  This should simplify URL matching as well.

jukkar replied:
This can be done by application which can fill the url array in desired order.

lpereira commented on subsys/net/lib/http/http_app_server.c:
```diff
@@ -0,0 +1,1085 @@
+	ctx->state = new_state;
+}
+
+static inline u16_t http_strlen(const char *str)
```
This is unused.

jukkar replied:
Indeed, good catch.

mike-scott commented:
I'll try and review the next version posted as I'm a bit buried atm.

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+
+This sample code generates HTTP 1.1 responses dynamically and does not serve
+content from a file system. The source code includes examples on how to write
+HTTP 1.1 responses: 200 OK, 400 Bad Request, 403 Forbidden, 404 Not Found and
```
add a comma before the "and" for the last item in this list.

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+This sample code generates HTTP 1.1 responses dynamically and does not serve
+content from a file system. The source code includes examples on how to write
+HTTP 1.1 responses: 200 OK, 400 Bad Request, 403 Forbidden, 404 Not Found and
+soft HTTP errors like 200 OK with a 404 Not Found HTML message.
```
s/like/such as/

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+
+The HTTP Server sample application for Zephyr implements a basic TCP server
+on top of the HTTP Server Library that is able to receive HTTP 1.1 requests,
+parse them and write back the responses.
```
add a comma before the "and" for the last item in this list.

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+it will respond with a soft HTTP 404 status code, i.e. an HTTP 200 OK status
+code with a 404 Not Found HTML body.
+
+The second line must be interpreted as follows: requests to /headers,
```
s/must be/is/

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+The second line must be interpreted as follows: requests to /headers,
+/headers/index.html and in general to /headers/xxx, will trigger the
+http_write_header_fields routine that prints the received HTTP
+Header Fields. In this case, "xxx" must be understood as any resource
```
s/must be understood as/is/

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+
+	index.html                                            [ <=> ]
+
+The inspection of the file index.html will show
```
s/The inspection of/Inspecting/

dbkinder reviewed (changes requested):
some suggested edits

nashif reviewed (changes requested):
Get rid of all those http_app_xxx just call it http, there is no reason why we should have app in the name

jukkar committed:
samples: net: HTTP client sample using net-app

This is a sample HTTP client application that uses new http API
instead of legacy one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: http: Deprecate old HTTP library

The old HTTP server and client library code is deprecated. The
new HTTP library will be based on net-app API code which requires
changes to function names and parameters that are not compatible
with old library.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: http: Remove deprecated HTTP client/server samples

As the old HTTP API is deprecated, the old samples can be removed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: https: mbedtls buffer length needs to be bigger for https

Mysterious TLS errors are printed if we try to work with too
small crypto buffer when https is enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented on subsys/net/lib/app/net_app.c:
```diff
@@ -509,11 +509,21 @@ struct net_context *select_client_ctx(struct net_app_ctx *ctx,
 		}
 	} else {
 		if (!dst) {
+			if (ctx->default_ctx->ctx &&
```
> It might happen in TCP client, that the TCP connection is terminated in which case net_context is freed. Check this and mark corresponding net_context inside net_app to NULL. This way there will be no issue to access already freed net_context.

This is very dangerous. What if it's freed and already reused?

Such condition should not happen, it's a bug and we should be fix it (which should be as simple as incrementing a refcount for it, or not so simple).

I'm writing in particular because I see some weird issues in MicroPython which could be explained by this issue to.

jukkar replied:
> What if it's freed and already reused?

That is why we mark the context as NULL here so that we do not reuse it in net-app.

I am not sure if adding refcount here would help as if TCP connection is closed, then we would need to inform the users of the net_context that it is closed. Probably using other flags in net_context we could figure this out in users of net_context.

pfalcon replied:
> That is why we mark the context as NULL here so that we do not reuse it in net-app.

What if it's reused by non-net-app?

> I am not sure if adding refcount here would help as if TCP connection is closed, then we would need to inform the users of the net_context that it is closed.

Well, (at the very least) we just need to make sure that the closed context is properly marked so (so any further operation on it will return appropriate error), but it's not freed until user closes ("puts") its own reference to it.

pfalcon reviewed (changes requested):
As discussed in RFC on the mailing list and in review comments to #980, I really think that the new API doesn't offer enough improvements over old one, and even calls for particular, questionable design choices in the underlying stack (e.g. #1330).

So, I'd like to propose to rework it. Let me start with a first: reworking a TX path for HTTP.

Proposal:

1. There should be a single function to handle sending of HTTP content, let it be called http_send_data() (this reuses the name of function used in this PR for completely different semantics, hope that's not too confusing with this note.
2. It's prototype to be int `http_send_data(struct http_ctx *ctx, const char *payload, size_t payload_len);` Note - no references to network packets or something.
3. `struct http_ctx` should have a member `struct net_pkt *tx_pkt`, initially set to NULL.
4. Algorithm of the function described in below points.
5. If `ctx->tx_pkt` is not allocated, allocate it.
6. Add to the packet whatever data of `payload` fits. (This will require #119).
7. If all user data handled, return.
8. When full, send the packet to the network and set `ctx->tx_pkt` to NULL.
9. Repeat from 5. for remaining user data.

All other functions can be implemented in terms of http_send_data().

In addition to http_send_data(), should have http_flush_data(), which will send whatever data has already buffered (useful e.g. for websockets), and http_finish_data() (HTTP 1.0 would close connection here).

The names are tentative. Possibly, http_send_data() won't be user-facing function (likely wrappers will be required to be user-facing to abstract away chunked encoding), so that function can be called differently and "http_send_data" reused for more user-facing func.

jukkar commented:
> even calls for particular, questionable design choices in the underlying stack (e.g. #1330).

I do not understand how #1330 is related to this PR, could you elaborate this claim?

pfalcon commented:
> I do not understand how #1330 is related to this PR, could you elaborate this claim?

This PR introduces such an HTTP API which then requires #1330 to handle some of the API "aspects". If the API is redesigned, going for solution like #1330 won't be required (with many issues raised with it).

jukkar commented:
> This PR introduces such an HTTP API which then requires #1330 to handle some of the API "aspects".

This is false, the API does not mandate the use of #1330 
I just wanted to simplify the **sample** application logic and avoid looping through data.

Anyway, I will investigate your proposal for API functions more next week.

pfalcon commented:
> Anyway, I will investigate your proposal for API functions more next week.

I appreciate that. I'll be happy to help with the implementation in any way I can.

jukkar commented:
Sent net-app changes separately in PR #4625

jukkar commented:
New version that is using the function proposed by @pfalcon

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,247 @@
+		}
+	}
+
+	ret = _http_add_chunk(ctx->pending, ctx->timeout,
```
Well, so the idea was that http_send_data() is the most basic, primitive function to send data over HTTP connection. All other sending is done solely via it. Sending a transfer-eccoding-chunked chunk is thus a higher-level operation, with implementation like:

~~~
add_chunk(ctx, buf, len)
{
  char sizebuf[10];
  int sz = sprintf(sizebuf, "%x\r\n", len);
  http_send_data(ctx, sizebuf, sz); 
  http_send_data(ctx, buf, len); 
  http_send_data(ctx, "\r\b", 2); 
}
~~~

jukkar replied:
I disagree here. IMHO the lowest function should be the one that is using the net_buf as a parameter. In this case it would be http_send_msg_raw(). If developer knows what he/she is doing, then that function could be used to send data.

pfalcon replied:
> If developer knows what he/she is doing, then that function could be used to send data.

That statement is always true, and I always agree with it. The point, in the paradigm I propose, such a function will never be needed by an end user, and the only use will be from `http_send_bytes()`, and all other funcs will use `http_send_bytes()` (and never `http_send_msg_raw()` directly). As long as we on the same line with that, we're talking about the same paradigm, otherwise, we're still talking about different things.

jukkar replied:
Sure, I agree.

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,247 @@
+	size_t added;
+
+	if (!ctx->pending) {
+		ctx->pending = net_app_get_net_pkt(&ctx->app_ctx, AF_UNSPEC,
```
Per above, only http_send_data() allocates ctx->pending, and it's the only function which calls net_pkt_append(). All other functions use http_send_data() for any data transfer.

pfalcon reviewed (changes requested):
Thanks for giving it a try. It's not yet quite there though, hopefully the inline comments elaborate/clarify that the idea is that `http_send_data()` is the only function which does buffer allocation/management, and all other functions just use its services. Again, perhaps the naming is what confusing - I proposed reusing older name for completely different semantics. Let it be called `http_send_bytes()`, to emphasize that it deals with just a stream of bytes, not higher levels protocol like chunked encoding, etc.

jukkar commented:
> Let it be called http_send_bytes(), to emphasize that it deals with just a stream of bytes, not higher levels protocol like chunked encoding, etc.

As that function might not send anything, that name is still not great. We could probably call it http_prepare_for_send() or similar.

jukkar commented:
Sent updated version according to comments. In this version I named the sending function to http_prepare_and_send() as it does exactly that. The send_bytes() and send_bytes_raw() are also good but sounded a bit off.

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,259 @@
+	size_t added;
+	int ret;
+
+again:
```
Why not explicit loop? ;-)

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,259 @@
+		goto again;
+	}
+
+	return added;
```
This value doesn't have much use, better to return 0, I guess.

jukkar replied:
Yes, now it does not make much sense indeed.

pfalcon commented on subsys/net/lib/http/http_app_client.c:
```diff
@@ -0,0 +1,720 @@
+		ctx->pending = NULL;
+	}
+
+	http_add_header(ctx, method, ctx->http.req.user_data);
```
Either error handling, or comment like "Here we expect that we can put a bunch of stuff into freshly allocated pkt (without it being sent, which can lead to I/O error), even though nothing in TCP spec says we could."

Yeah, that's why they invented exception handling... I personally try to wrap that into macros whenever possible, IMHO it's very cumbersome to do it otherwise.

pfalcon replied:
Note that (besides exception handling) another way to deal with it is to have an explicit buffer to construct request (not relying on the fact that we can allocate given amount of data in a pkt, because we can't - MSS can be negotiated to be 1).

pfalcon commented on subsys/net/lib/http/http_app_client.c:
```diff
@@ -0,0 +1,720 @@
+			goto out;
+		}
+
+		for (i = 0; i < req->payload_size;) {
```
This loop should not be needed (chunked encoding itself doesn't have limitations on chunk size, right?). http_send_chunk() uses http_prepare_and_send(), and that guarantees that it sends all the data (unless error occurs). That's the beauty of it (the downside is the need to check each call for error - but that's more the errors to blame, not the approach ;-).

pfalcon replied:
But it's not needed ;-). Let's spend those code bytes on adding error checking consistently everywhere.

jukkar replied:
The API has now http_send_chunk() which adds a chunk. If we want to have only one chunk in payload regardless of size, then this API is not useful and the caller would need to know the total size of the sent data in advance. So I propose that we go with this way (multiple chunks) atm.

pfalcon replied:
Well, we actually should rework HTTP request API to allow to do a streaming POST for example ;-). I don't raise that now, because there're enough improvements for this patch already. OTOH, think about that now will allow to deprecate less API calls later. So, let me know if you'd like to discuss that or will have thought about that yourself.


Otherwise, we know that http_send_chunk(..., len) will return len (that's the API contract), so the loop will run only once. We can leave it until next time of course.

pfalcon reviewed (commented):
That's it! Added few comments still.

pfalcon commented:
With the core send processing settled, let's think what's the purpose of send callback. I see only one purpose: to let user be able to show some progress indicator. And with the current implementation, that would be more like rotating bar, not even a percentage being sent. Is that's the intended usecase? And in either case, let's think what's the purpose of user_send_data param. There's no guarantee that call to any send*/add* function will lead to the callback called. Then why do we pass that data around? Let user set it once in the context (and update as they see fit), and use that instead.

jukkar commented:
> With the core send processing settled, let's think what's the purpose of send callback. I see only one purpose: to let user be able to show some progress indicator. And with the current implementation, that would be more like rotating bar, not even a percentage being sent. Is that's the intended usecase? And in either case, let's think what's the purpose of user_send_data param.

For server connection, the send callback might have limited usage. But in the client side it could be used for various purposes. And as we have shared functions between client and server I think we need to have the sent callback. User does not need to use it so it does not really harm to have the cb and related user_data around.

>  There's no guarantee that call to any send*/add* function will lead to the callback called. Then why do we pass that data around? Let user set it once in the context (and update as they see fit), and use that instead.

The sent callback will be called after the data is sent, so at some point of the call flow the callback will be called.

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,259 @@
+	snprintk(chunk_header, sizeof(chunk_header), "%x" HTTP_CRLF,
+		 (unsigned int)len);
+
+	http_prepare_and_send(ctx, chunk_header, strlen(chunk_header),
```
Error checking.

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,259 @@
+
+	if (len) {
+		len = http_prepare_and_send(ctx, buf, len, user_send_data);
+		if (len == 0) {
```
This if should be error checking instead.

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,259 @@
+
+	ret = http_prepare_and_send(ctx, HTTP_CRLF, sizeof(HTTP_CRLF),
+				    user_send_data);
+	if (ret == 0) {
```
This if should be error checking instead.

pfalcon commented:
> The sent callback will be called after the data is sent, so at some point of the call flow the callback will be called.

Yes, but for 2 calls, `send(..., user_data1)` and `send(..., user_data2)`, there's no guarantee that callback will be called with user_data1 or user_data2 (or both at all). That's the motivation for the proposal to make user data more static. (That can be seen also as requesting the growth of ctx structure to hold that, if you don't want to do it because of that, just comment, and then sounds good).

jukkar commented:
> Yes, but for 2 calls, send(..., user_data1) and send(..., user_data2), there's no guarantee that callback will be called with user_data1 or user_data2 (or both at all). That's the motivation for the proposal to make user data more static. (That can be seen also as requesting the growth of ctx structure to hold that, if you don't want to do it because of that, just comment, and then sounds good).

Actually having different user_data for each request is possible (supported by net_context). I noticed an issue with client API regarding this so will fix it.

pfalcon commented on include/net/http_app.h:
```diff
@@ -0,0 +1,1079 @@
+					    const char *host,
+					    const char *extra_header_fields,
+					    const char *content_type,
+					    const char *payload,
```
(if you want to continue POST discussion). Here you effectively assume that POST payload is textual data. Why so? What's so wrong with posting binary file for example? (Yeah, adding yet another param will make signature even more cumbersome, but "stray" params there are `http_response_cb_t cb`, `user_data` (perhaps also response_buf - just malloc() it away, lol). Perhaps, we don't need these wrappers at all, and users should use `struct http_request` directly.

pfalcon commented:
> Actually having different user_data for each request is possible (supported by net_context). I noticed an issue with client API regarding this so will fix it.

For each request is possible and makes sense. What the current API is trying to do is literally different user_data for each byte (sequences of bytes), which won't work that way.

jukkar commented:
> What the current API is trying to do is literally different user_data for each byte (sequences of bytes), which won't work that way.

The sent callback with user specified user_data is only called when we actually sent things. Of course the parameter is given to every invocation of the function that might sent data as we do not know when the data is actually sent.

pfalcon commented:
> The  sent callback with user specified user_data is only called when we  actually sent things. Of course the parameter is given to every  invocation of the function that might sent data as we do not know when  the data is actually sent.

That "of course" will confuse someone sooner or later. A simple solution was proposed: user data should be part of (per-request) context, to give user idea it's actually per request, not per something else. (Perhaps, it can/should be the same user data for all callbacks). If that doesn't ring a bell so far, can be revisited later.

jukkar commented:
New version with sanity fixes

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,264 @@
+	return 0;
+}
+
+int http_prepare_and_send(struct http_ctx *ctx,
```
Can you please put  http_prepare_and_send() above _http_add_chunk() (i.e. to form a logical hierarchy)?

jukkar replied:
sure

pfalcon commented on subsys/net/lib/http/http_app.c:
```diff
@@ -0,0 +1,264 @@
+		    size_t payload_len,
+		    void *user_send_data)
+{
+	return _http_add_chunk(ctx, payload, payload_len, user_send_data);
```
Do we need _http_add_chunk() at all, can you put its body here?

jukkar replied:
Yep, that one is unnecessary func.

pfalcon commented on subsys/net/lib/http/http_app_client.c:
```diff
@@ -0,0 +1,738 @@
+	const char *method = http_method_str(req->method);
+	int ret;
+
+	if (ctx->pending) {
```
I'd abstracted this away into a func, the q is how to call it again...

pfalcon commented:
(This is a bit big patch for review after all, would be nice to have summary of changes in new versions.)

So, I don't see http_client_send_post_req() in the code, only in the header, was anything done to it, or am I missing something?

jukkar commented:
> So, I don't see http_client_send_post_req() in the code, only in the header, was anything done to it, or am I missing something?

Hmm, I do not recall any comment regarding that API, you want that changed too?

pfalcon commented:
> >   So, I don't see http_client_send_post_req() in the code, only in the header, was anything done to it, or am I missing something?

> Hmm, I do not recall any comment regarding that API, you want that changed too?


The comment is here: https://github.com/zephyrproject-rtos/zephyr/pull/4243#discussion_r148028982 , I suggest that it's not general enough to not take the payload size (rules out use with binary payloads).

jukkar commented:
> The comment is here: #4243 (comment) , I suggest that it's not general enough to not take the payload size (rules out use with binary payloads).

Indeed. I will remove the POST helper, one is able to send a POST request without it anyway.

pfalcon commented:
Makes sense, thanks!

jukkar commented:
Changes:
* fix merge conflicts
* fix sanity errors
* removed POST req client helper
* moved functions around in http_app.c
* combined two chunk send functions together

pfalcon reviewed (approved):
Ok, I think this is a very good progress in the right direction.  My general suggestion is that we treated such work as an iterative process, and didn't try to replace an older API version at once. For example, I may imagine, further changes/fixes will be required as this API gets more use.

But for iteration one it's really good, +1.

jukkar commented:
@dbkinder the documentation checker fails with this message
```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/net/http_app.h:1040: warning: reached end of file while inside a verbatim block!
The command that should end the block seems to be missing!

/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/net/http_app.h:1040: warning: end of file while inside a group
```
I have spent several hours trying to understand what goes wrong but have failed so far. The header file contains proper @{ and @} markers so it baffles me what is the root cause of this warning. And doxygen does not tell what is the actual problem it encountered. Could you help here to solve this issue?

jukkar commented:
@nashif Ping, does this look something that is acceptable to you? I deprecated the old API, but if you want I can certainly remove it alltogether.

dbkinder commented:
@jukkar As a temporary workaround, I found that changing the end comment to add an @endverbatim directive makes the doxygen warning go away.  Maybe put that in for now so you can get this PR merged and I'll continue tracking down what's going on:

     /**
      * @}
      * @endverbatim
      */

dbkinder commented:
@jukkar  It took quite a while but I found the culprit.  I ended up just deleting chunks of lines, running doxygen on just this file, and finding the line(s) causing the problem.  Doxygen doesn't like to see \r in a doxygen comment.  You've got a few in here as  \r\n that I deleted (leaving just the \n) the the verbatim block and end of file in a group warning went away, yea!  Maybe replace use of \r\n in doxygen comments as <return><newline>

That then exposed some errors because you missed the documentation for a couple of function parameters:

http_app.h:967: warning: The following parameters of http_add_header(struct http_ctx *ctx, const char *http_header_field, void *user_send_data) are not documented:
  parameter 'user_send_data'

http_app.h:983: warning: The following parameters of http_add_header_field(struct http_ctx *ctx, const char *name, const char *value, void *user_send_data) are not documented:
  parameter 'user_send_data'

dbkinder commented on include/net/http_app.h:
```diff
@@ -0,0 +1,1039 @@
+ * @param host Host field in HTTP header. If set to NULL, then server
+ * name is used.
+ * @param extra_header_fields Any extra header fields that caller wants
+ * to add. This can be set to NULL. The format is "name: value\r\n"
```
Can't use ``\r`` in a doxygen comment.  Recommend something like this instead:

     The format is "name: value<HTTP_CRLF>"

jukkar replied:
Thanks David! Really peculiar issue indeed, especially as the warning did not match the root cause.

pfalcon replied:
> Can't use \r in a doxygen comment. 

Maybe it should be double-quoted, like `\\r`? @jukkar , please try that first, because having `\r\n` is clearer than `<HTTP_CRLF>`.

dbkinder reviewed (changes requested):
\r in doxygen comments can cause odd "end of file while inside a verbatim block" warnings

jukkar commented:
Changed the documentation so that sanity check passes without warnings.

dbkinder commented on doc/zephyr.doxyfile:
```diff
@@ -113,7 +113,7 @@ INPUT                  = \
 FILE_PATTERNS          = "*.c" "*.h" "*.S"
 RECURSIVE              = YES
-EXCLUDE                = ../include/spi_legacy.h
+EXCLUDE                = ../include/spi_legacy.h ../include/net/http_legacy.h
```
If we're starting a trend to ignore *_legacy.h files, better would be to delete these individual files listed here and change the line below to be:

     EXCLUDE_PATTERNS = *_legacy.h

jukkar replied:
I am ok with this. Perhaps this could be a separate patch and not part of this set.

dbkinder replied:
Ok. I’ll submit a separate patch for ignoring all _legacy.h files.

dbkinder commented on include/net/http_app.h:
```diff
@@ -0,0 +1,1090 @@
+ * @param host Host field in HTTP header. If set to NULL, then server
+ * name is used.
+ * @param extra_header_fields Any extra header fields that caller wants
+ * to add. This can be set to NULL. The format is "name: value<HTTP_CRLF>"
```
I've verified that using ``\\r\\n`` works in doxygen, so if you'd like to use that, I think the output would be nicer, though the way you've written this works for me since it references the #define above.

jukkar replied:
Ok, I think too that \\r\\n looks better in the documentation. I will send a new version that changes this.

jukkar commented:
Updated the documentation and changed <HTTP_CRLF> to `\\r\\n`

jukkar commented:
Fixed the http_url_flags enum values in header file.
